[Tey Hao Ze] iP

Haozeee opened the pull request:
# Duke

> The key to success is having a good task manager. - Unknown

Duke frees your mind from having to remember all your trivial tasks. It's,
* Text-based
* Lightweight
* Easy to learn
* ~~FAST~~ SUPER FAST to use
* **COMPLETELY FREE OF CHARGE!!!**

Here are some of the features of Duke:
- [x] Store different types of tasks such as deadlines or events.
- [x] Search for tasks based on keywords
- [x] Features to manage your task list easily
- [ ] Search for tasks by date (Coming soon)
- [x] Cool GUI (Coming soon)

If this sounds like something you want, all you need to do is,
1. Download it from [here](https://github.com/Haozeee/ip).
2. Double-click it.
3. Add some tasks.
4. Let it manage your tasks for you :satisfied:

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method.
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

Haozeee committed:
Level-1

Haozeee committed:
Level-2

Haozeee committed:
Level-3

Haozeee committed:
level-4

Haozeee committed:
level-4

Haozeee committed:
A-TextUiTesting

Haozeee committed:
Level-5

Haozeee committed:
Level-6

Haozeee committed:
A-Enums

Haozeee committed:
Add a few javadocs comments

Haozeee committed:
Add save feature

Haozeee committed:
Add date and time feature.

Haozeee committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

Haozeee committed:
Add Command classes

A Command interface has been added to represent commands given by the user
to Duke.

Haozeee committed:
Add packaging to project

The java files are now separated into different packages.
The following packages were added:
*commands
*duke
*exceptions
*tasks

Haozeee committed:
Add unit tests with JUnit

Unit tests were added for the following classes:
*AddDeadlineCommand
*AddEventCommand
*DateParser
*DukeDate
*Task

Haozeee committed:
Add JAR

Haozeee committed:
Update test cases

Haozeee committed:
Add JavaDocs

Haozeee committed:
Follow Java coding standards

Haozeee committed:
Add find task functionality

Haozeee committed:
Merge branch 'branch-A-CodingStandard'

Haozeee committed:
Merge branch 'branch-Level-9'

Haozeee committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

Haozeee committed:
Add Gradle support

Haozeee committed:
Add CheckStyle config files

Haozeee committed:
Add a GUI to Duke

Haozeee committed:
Add varargs feature

Users can now mark multiple tasks as completed or delete multiple tasks in one command by giving a list of index.

Previously if a user wanted to mark task 1, 2, 3 and 4 as done, the user had to enter 4 separate commands.
Now, the user can just enter done 1 2 3 4 in order to mark the 4 tasks as completed.

Haozeee committed:
Remove previously used command line functions

Haozeee committed:
Add minor variable name changes

Haozeee committed:
Merge branch 'branch-A-CheckStyle'

Haozeee committed:
Add Java check style functionality

Haozeee committed:
Improve GUI

Chats now have colours and better fonts.

Haozeee committed:
Improve GUI

peilinye commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,55 @@
+        this.type = typeToAdd;
+    }
+
+    public abstract boolean verifyAddCommand(String input);
```
Perhaps a header comment could be added for this public method and the other 2 below?

peilinye commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,55 @@
+ * An abstract class that represents commands given by the
+ * user that involves adding a task to the taskList.
+ */
+public abstract class AddCommand extends Command {
```
I like how you made the `AddCommand` class abstract and used `AddEventCommand`, `AddDeadlineCommand`, etc.

peilinye commented on src/main/java/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,64 @@
+        this.taskList = taskList;
+    }
+
+    @Override
```
Perhaps a header comment here to briefly describe what this method does would be good? Could add in header comments on other public methods as well.

peilinye commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+public class TaskList {
+
+    /** An arraylist that contains the tasks set by the user. */
+    private final ArrayList<Task> taskArrayList = new ArrayList<>();
```
Perhaps a more intuitive variable name here? such as 'tasks'

peilinye commented on src/main/java/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,64 @@
+        }
+        return true;
+    }
+}
```
I like how the different add commands parse the user input and check for input errors, the code is clean this way. :)

peilinye commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+    /** An arraylist that contains the tasks set by the user. */
+    private final ArrayList<Task> taskArrayList = new ArrayList<>();
+    /** The number of uncompleted tasks in the task list */
+    private int uncompletedTasks = 0;
```
Perhaps a more intuitive variable name here such as 'numOfUncompletedTasks'? Same for 'totalTasks'

peilinye reviewed (commented):
Overall, your code is neat and easy to read. There were not many coding standard violations, but you could consider adding more header comments for public methods.

dhshah1 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,119 @@
+            return "[T]";
+        } else if (this.type.equals(Type.EVENT)) {
+            return "[E]";
+        } else {
```
Very small suggestion: Perhaps it would be better to still keep an else if clause here, and then handle the else case when it doesn't match any type as well. It's unlikely that this will be an issue but maybe its better to be defensive here in case in the future there are more task types and this method is not updated by mistake

```suggestion
        } else if (this.type.equals(Type.DEADLINE) {
```

dhshah1 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Marks a task as completed.
+     */
+    public void setAsFinished() {
```
Perhaps this method could be named as setAsDone or markAsDone? Just so that the terminology is consistent instead of having "finished" and "done".

dhshah1 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,119 @@
+            loadedTask = Todo.newTodoTask(taskName);
+        } else if (typeOfTask.equals("D")) {
+            loadedTask = new Deadline(taskName, DukeDate.getDukeDateFromType(strComponents[4].strip(), dukeDateType));
+        } else {
```
Same as above, perhaps to be defensive this can also be changed to an else if clause

dhshah1 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,119 @@
+     * @param saveString The String that represents the saved task.
+     * @return The task converted from saveString.
+     */
+    public static Task stringToTask(String saveString) {
```
Maybe this method could be renamed to something like storageStringToTask, just to make it clear that its specifically talking about parsing the storageString format

dhshah1 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     *              So the index 1 represents the first task in the taskList ArrayList.
+     * @return The message that should be shown to the user after a task has been marked as done.
+     */
+    public String markTaskAsCompleted(int index) {
```
Perhaps the terminology for a completed task could be standardised? Currently there's a few terms used to refer to this: "finished", "done", "completed".

dhshah1 reviewed (commented):
Overall, I felt the code was really well written and organised. Honestly, I learnt quite a lot from going through the code, so there weren't really any major coding quality issues I picked up on. Just left some minor comments below for your consideration.

Haozeee committed:
Update code after peer review

Peer review last week indicated some minor coding style violations in the codebase. This commit corrected some of those coding violations that have been pointed out.

Haozeee committed:
Add assertions

Assertions were added to the codebase to ensure code correctness during runtime.

The following are the classes and the methods where assertions were added:
* duke.Parser: parseUserInput
* duke.Storage: saveTaskList
* javafx.MainWindow: setMainWindow
* tasks.TaskList: addTask, removeTask, markTaskAsDone

Haozeee committed:
Improve code quality

The code was updated in order to comply with Java Code quality.

One of the major changes was to edit some of the if-else branches where the else branch was used to specify the final option. This may lead to errors in the future.

The code has been changed such that most of else in the if-else branch means everything else and not the final option.

Haozeee committed:
Merge pull request #1 from Haozeee/branch-A-Assertions

Add assertions

Haozeee committed:
Merge branch 'master' of https://github.com/Haozeee/ip

* 'master' of https://github.com/Haozeee/ip:
  Add assertions

Haozeee committed:
Merge branch 'master' into branch-A-CodeQuality

Haozeee committed:
Merge pull request #2 from Haozeee/branch-A-CodeQuality

Improve code quality

Some of the if-else branches has been changed so that the else means "everything else" rather than just the final option. This will reduce the possibility of errors in the future.

Haozeee committed:
Merge branch 'A-CodeQuality' to master

The master branch has been updated to include changes made on the 'A-Code-Quality' branch. The code now complies with the Java coding standard. More information on the changes can be found in the branch's commit messages.

Haozeee committed:
Resolve an error from merging

Haozeee committed:
Update javadocs

Haozeee committed:
Comply with CheckStyle

Haozeee committed:
Comply with Java code quality

Made minor changes to improve the readability of the code and make it less error prone.

Also, more assertions were added.

Haozeee committed:
Add undo extension

Added the feature for users to undo any previous changes to the task list. Users can now undo up their commands up to 10 times.

Haozeee committed:
Add Undo extension

Haozeee committed:
Revert "Add undo extension"

This reverts commit aac23579354936dbe7a705d5a6c7b0a70301cdce.

Haozeee committed:
Change implementation of undo

Previously, undo was implemented by making a copy of the taskList as it changes.

Now, undo is implemented by storing the opposite command in each command that can be undone. This means we only need to store a copy of the command rather than the entire state. This reduces the memory usage of the program.

Haozeee committed:
Improve code quality

Haozeee committed:
Improve GUI

Added improvements such as images to make the GUI more attractive.

Haozeee committed:
Update README.md

Haozeee committed:
Update GUI

Haozeee committed:
Merge branch 'master' of https://github.com/Haozeee/ip

* 'master' of https://github.com/Haozeee/ip:
  Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update gitIgnore file

Haozeee committed:
Update GUI

Haozeee committed:
Update README.md

Haozeee committed:
Add images

Haozeee committed:
Add images

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Add images

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
change font

Haozeee committed:
Correct minor code quality issues

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Remove html tags

Haozeee committed:
Update README.md

Haozeee committed:
Update README.md

Haozeee committed:
Correct minor check style violation

Haozeee committed:
Update README.md
